core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -925,51 +933,195 @@ public boolean isHashValid(String hash) {
         }
     }
 
-    /* package */ static final JBCryptEncoder JBCRYPT_ENCODER = new JBCryptEncoder();
+    public static class PBKDF2PasswordEncoder implements PasswordHashEncoder {
+
+        public static final int HASH_LENGTH = 8;
+        public static final int ZERO = 0;
+        public static final int HEX_TWO = 2;
+
+        public static final int RADIX_LENGTH = 16;
+        public static final String STRING_SEPARATION = ":";
+
+        public static final int KEY_LENGTH = 512;
+        public static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA512";
+
+        @Override
+        public String encode(CharSequence rawPassword) {
+            return generatePasswordHashWithPBKDF2(rawPassword);
+        }
+
+        @Override
+        public boolean matches(CharSequence rawPassword, String encodedPassword) {
+                return validatePassword(rawPassword.toString(), encodedPassword);
+        }
+
+        @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Accessible via System Groovy Scripts")
+        @Restricted(NoExternalUse.class)
+        private static int ITERATION_COUNT = SystemProperties.getInteger(HudsonPrivateSecurityRealm.class.getName() + ".ITERATION_COUNT", 10);
+
+        private static String generatePasswordHashWithPBKDF2(CharSequence password)  {
+            int iterations = (int) Math.pow(2, ITERATION_COUNT);
```

Comment:
As per latest guidelines from the client we go along with new format where iteration count is given explicite not as a power of 2